src/dmd/backend/dvec.d

```diff
@@ -0,0 +1,605 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (c) 2013-2018 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/backend/dvec.d, backend/dvec.d)
+ */
+
+module dvec;
```

Comment:
Should be module dmd.backend.dvec, otherwise separate compilation fails.

Reply:
Aye, there's the rub. I cannot make this work for both dmc and dmd, because they use separate directory structures. Upgrading the boot compiler to 2.079 is required. Unfortunately, this has been blocked for a year now, and is currently blocked by:
https://github.com/dlang/dmd/pull/8249
https://github.com/dlang/dmd/pull/8112

Reply:
I haven't yet heard a convincing reason why dmc and dmd should not share a common directory structure, e.g. use module name `tk.dvec` and put it into the respective folder. Not using package names is unlikely to be a good idea for shared modules.

Reply:
The reason is:

1. It's a lot of nuisance work for me to crowbar dmc into a totally different directory structure - messing up the git history, changing all the imports, etc.
2. I'm the one doing the work, and this is the way I want to do the work, because mashing it into multiple directories is unproductive for me.
3. All the blocking items for making this work are inconsequential and do not negatively impact anything else.
4. All the blocking items are things we need to do anyway (like support the -mv command). I can hardly be the only one who wants to share the same modules among multiple projects.
5. assert() support in 2.068 is so heavy it is disabled in dmd, with various negative consequences that show up now and then. The back end relies heavily on asserts being ON. 2.079 has much better support for lightweight assert().

Reply:
> It's a lot of nuisance work for me to crowbar dmc into a totally different directory structure - messing up the git history, changing all the imports, etc.

You'll have to change .c files to .d in your build system anyway. Is it any worse to prepend `tk/` or similar, too?

>     I'm the one doing the work, and this is the way I want to do the work, because mashing it into multiple directories is unproductive for me.

Valid point. Don't consider my comments as blockers.

>     All the blocking items for making this work are inconsequential and do not negatively impact anything else.
>     All the blocking items are things we need to do anyway (like support the -mv command). I can hardly be the only one who wants to share the same modules among multiple projects.

That's what packages are designed for. Adding -mv showed that you seem to think that it has failed, but I don't see how. BTW: if you don't want to use package names anyway, -I is good enough. IMO the existence of -mv just adds another point of failure and confusion.

>     assert() support in 2.068 is so heavy it is disabled in dmd, with various negative consequences that show up now and then. The back end relies heavily on asserts being ON. 2.079 has much better support for lightweight assert().

AFAICT the difference in code size/performance is negligible between C and D asserts (just strings instead of raw pointers). I just tested removing C asserts from the backend: compiling release-dmd itself with a VC built 64-bit version took 110s instead of 114s with asserts enabled (dmc built dmd2.080 took 137s). IIRC the change in the frontend was in the same ballpark.

Reply:
See https://github.com/dlang/dmd/pull/5716 for more discussion on the effects of `assert` on the compiler.

Reply:
> Is it any worse

Yes. I generally try to make the first .d version close enough that git will show the diffs.

> but I don't see how

For one thing, the unzip utility will no longer unzip subdirectories that are not in the cwd. So if you want to zip up the project, you're hosed unless they share subdirectories. But dmd and dmc use different hierarchies.

I understand you don't like `-mv`. Nobody is making anyone use it, and so it doesn't negatively impact them. It's also TRIVIAL. I simply cannot understand the resistance to it. (I've implemented a lot of stuff that I don't personally find useful, but others do.) It's a heluva lot more work for me to disrupt all the layouts on the various machines I use, and the other git repositories that dmc resides in, than just add `-mv=` to the command line, and I'm done.

Your proposal is like what happens when your credit card gets a fraud alert. You then have to notify and change the number with every vendor you've got an account with, and then for months afterwards you get dinged on the ones you overlooked. I've suffered through that many times, nobody does that for fun if they don't have to.

It's not just me, it's not a nice thing to inflict on our users, either.

I've been repeating these arguments for a year. I could have had the back end converted a year ago. Instead I spend time on this back and forth over a trivial blocker.

Reply:
> sed scripts

Adding more technical debt is to be avoided, not embraced. Never mind that sed isn't a standard tool on Windows, so I'd have to write a program to do that, and check it into the repositories. I know what I think when I see such things in other projects.

> trivial

Yes, I could write a script to copy directory structures around before each compile. Is that something you'd want to do for your project? I prefer to avoid rube goldberg build setups. (Heaven knows the build system is fragile enough as it is.)

> pretending

I'm used to doing ugly things to get things to work when I have no choice. I've even resorted to writing programs to patch object files. But there's no reason to do that here. The advantage to owning the tool chain is the fix can be put in the right spot.

> Your word still carries weight here.

Then please pull the blocking PRs.

Reply:
I had my own non-standard `assert` for years in the back end. I finally removed it along with a lot of other garbage I used to think was a good idea.

Reply:
> One is failing,

Because it's blocked by the other.

> and @kinke seems to have a valid suggestion for the other one.

Regardless of the merit of his suggestion, it requires delaying for multiple release cycles.